components/suggest/src/store.rs

```diff
@@ -194,6 +189,20 @@ impl SuggestStore {
         self.inner.query(query)
     }
 
+    /// Dismiss a suggestion
+    ///
+    /// Dismissed suggestions will not be returned again
+    #[handle_error(Error)]
+    pub fn dismiss_suggestion(&self, suggestion: Suggestion) -> SuggestApiResult<()> {
```

Comment:
Hi @0c0w3, FYI, this is the dismissal API exposed by the component. It will take a `Suggestion` as the only argument. Any concerns about that?

Reply:
Thinking about it a bit more (sorry I didn't think of this earlier!), the one downside I can think of is that the caller needs to hold on to the entire `Suggestion` for as long as it's visible.

We can do that on Android and iOS—there's a [`metadata` map](https://github.com/mozilla-mobile/firefox-android/blob/8a69a9a83cbb86132d68f81e0e7f509a146c167c/android-components/components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt#L105-L106) we can use to smuggle it on Android, and we can [add a field on iOS](https://github.com/mozilla-mobile/firefox-ios/blob/0194fca784ee190138615652439d9fb4ef9cee4e/firefox-ios/Storage/Rust/RustFirefoxSuggestion.swift#L24-L28)...

_But_ since the suggestion carries the raw `icon` data, we're duplicating the icon data in memory: once for the buffer in the `Suggestion`, and once for the decoded [`android.graphics.Bitmap`](https://github.com/mozilla-mobile/firefox-android/blob/8a69a9a83cbb86132d68f81e0e7f509a146c167c/android-components/components/feature/fxsuggest/src/main/java/mozilla/components/feature/fxsuggest/FxSuggestSuggestionProvider.kt#L126) / [`UIImage`](https://github.com/mozilla-mobile/firefox-ios/blob/0194fca784ee190138615652439d9fb4ef9cee4e/firefox-ios/Storage/Rust/RustFirefoxSuggestion.swift#L67) instance. I think we'll also incur a third copy of that icon data when we pass it back over the FFI to `dismiss_suggestion`.

This would be a backward-incompatible change, but I wonder if it's worth making `Url` a dictionary or an opaque `interface` type. Alternatively, maybe we could add a `dismissal_id` field—either a string, or an opaque `interface`—that carries the dismissal URL.

Reply:
Right, I was initially thinking about the [special handling](https://searchfox.org/mozilla-central/source/browser/components/urlbar/private/SuggestBackendRust.sys.mjs#385-400) we did for desktop and wondered if it's possible to restore the `Suggestion` object and pass it back to the component with all needed information, especially, the enum variant type. Lina's comment makes this even more questionable.

How about we just take a `String` instead as the only input for this API? Or, if we expect to include other fields for a dismissal in the future, we can add an interface `Dismiss` with a string `url` field for now and use this to capture the context for this API.

Reply:
I like the idea of having the consumer send us a URL rather than the full Suggestion.

I think now that https://bugzilla.mozilla.org/show_bug.cgi?id=1875848 has been implemented, we can do a type check to figure out the variant when we try to extract the URL.  This still means some duplicated code, but it's not that bad.

Going forward, I think we could probably improve the API shape.  Now that https://github.com/mozilla/uniffi-rs/pull/1869 is merged, it should be possible to define enums that store objects which gives us a tone of options.  Maybe the `icon` field could be an object.  I think that would allow us to pass Suggestion instances around efficiently (assuming that's the only field that's storing a large amount of data).